[versionlock] update of the versionlock python script

vlours opened the pull request:
- To allow the check and diff options
- To include yum and dnf management
- To include add/exclude/delete/clear action in versionlock plugin

Commit:
[versionlock] update of the versionlock python script
- To allow the check and diff options
- To include yum and dnf management
- To include add/exclude/delete/clear action in versionlock plugin

vlours commented:
Hi @florianpaulhoberg,
I found your versionlock Ansible plugin and it was great but missed few stuffs.
I made few changes to allow the script to support the '--diff' and '--check' Ansible's options.
I also make sure that the script can support the versionlock options: add/excluded/delete/clear
And finally, I've included the dnf compatibility for the script.
The default package manager can be defined bases on the script name:
- yum_versionlock.py will manage yum by default
- dnf_versionlock.py will manage dnf by default.

I hope you will like this new version.
I also found the issue about versionlock in Ansible github.
I will raise a new one if you are happy with this new script.

I wish you a very Happy Holiday season.
Cheers

florianpaulhoberg commented:
Hi @vlours and a happy new year,

sorry for my late response, I've been on vacation and thanks for your PR. On the first view this looks brilliant, I'll give it a try and merge it.

Thanks!
Florian

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,265 @@
         argument_spec=dict(
-            state=dict(required=True, type='str'),
-            package=dict(required=True, type='str'),
+            state=dict(required=True, type='str', choices=['present', 'excluded', 'absent', 'clear']),
```
To match DOC (line 66) "excluded" should be set to "exclude" to prevent missleading error messages:
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "value of state must be one of: present, excluded, absent, clear, got: exclude"}

vlours replied:
Actually, the code is working with excluded, I will update the doc to match it.

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,265 @@
+        diff=dict(),
+    )
+
+    # Check for the versionlock config file and if plugins is enabled
```
Module will fail if there isn't any of them present:
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: IOError: [Errno 2] Datei oder Verzeichnis nicht gefunden: '/etc/yum/pluginconf.d/versionlock.conf'

vlours replied:
I will add a "try" to get a better behaviour when the file is absent.

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,265 @@
-        module.exit_json(changed=True, meta=response)
+    # Raw version will be used for the --diff option
+    versionlock_packages_raw = fct_versionlock(module, repo_mgr, 'list')
+    regex_exclude = re.compile(r'^Last metadata expiration check:.*')
```
Works only on EN/US locales.

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,265 @@
+        return out.splitlines()
     else:
-        module.fail_json(msg="Error: " + str(err) + str(out))
+        if rc_code == 1 and str(err) == 'Error: No matching Packages to list\n':
```
On non us locales:
fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Unable to collect yum list installed : Fehler: Keine übereinstimmenden Pakete zum Auflisten\n - "}

vlours replied:
Fixed in commit 4f0d34c290a259a437fa9b7605093d0faa6aeb41

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,265 @@
+        # Check if package is defined
+        if package == None:
+            module.fail_json(msg="Error: 'package' option is required (except for 'clear' state)")
+        List_available = get_packages(module, repo_mgr, 'available', package)
```
To keep the vars equal, maybe everything should be kept in lower case.

gyptazy reviewed (changes requested):
Hi @vlours,
thanks for you work and it seems to be cool on system based on EN/US locales but we should make sure to make this work on other locales, too.

Thanks,
Florian

vlours commented:
Hi Florian,

Happy new year to you too.

I check the code and try to get it work for the non us locales, or maybe just force the language to en_US as the module should only return the command's status in debug mode.

Did you had a chance to test it with en_US locales to check the results ?

Cheers,
Vincent

florianpaulhoberg commented:
Hi Vincent,

thanks, I already validated that it works with 'LANG=en_US.UTF-8', that was just meant as a hint for other locales. In the wild, there're too many different usecases and environments to fit. So there should be a more generic solution than just forcing everyone to use en_* locales. Workaround could be overwriting LC_ALL (should work here since yum/dns gets invoked).

Regards,
Florian

Commit:
[versionlock] update of the versionlock python script
- Include additional check for yum/dnf usage
- Update the documentation to ensure 'exclude' is replaced by 'excluded' for the 'state' parameter.
- Fix issue when the config file was missing
- Replace uppercase letter in variables.
- Set the locale to 'C' to get the yum result matching the regex expressions

vlours commented:
Hi Florian,

I've made some changes as suggested.
I've also included a new test to match RHEL8/Centos8, which are no longer using yum_versionlock, but only the dnf version.

I tried to test the language issue on my side, unfortunately, I was not able to get the language set at the system level in my docker containers.
Could you please test it in your side and let me know if the `locale.setlocale(locale.LC_ALL, 'C')` on line 81 does the job ?

Cheers

florianpaulhoberg commented:
Good morning @vlours,

looks good, unfortunately this way seems not to work for locales here (I'm surprised, too). I'd suggest to update the environment vars within the run_command. As an exmaple (I could already successfully validate this way) you could use:

# Global VAR:
ENV_LOCALE = {'LC_ALL': 'C'}

def get_packages(module, repo_mgr, list_type, package):
    """ Get an overview of all packages available/installed """
    rc_code, out, err = module.run_command("/usr/bin/%s -q list %s %s"
                                           % (repo_mgr, list_type, package), environ_update=ENV_LOCALE)
[...]

I think that's the fastest way to use it without rewriting the whole module as a class or to assign it to every func that uses run_command.

Regards,
Florian

Commit:
[versionlock] update of the versionlock python script
- Set the 'environ_update' for the commands to force the language to 'C'
- Affine Regex Expressions
- Reformat some warning/fail messages
- Create an alternate dictionnay to store the 'debug' content, and merge it at the end.
- Recreate the 'versionlock_packages_raw' from the array after cleaned up undesired reposync messages

vlours commented:
Hi @florianpaulhoberg,

Just made the suggested changes in the script.
I've also refined the Regex expressions, updated some warning/fail messages and also made few changes in the debug results.

I hope it will work for you now.

Cheers,
Vincent

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,283 @@
+    package_regex = re.sub('\*', '.*', package)
+    regex_search = re.compile('!?(\d+:)?%s-(\d+:)?[.\w]+\.\*' %package_regex)
+    for locked in versionlock_packages:
+        #versionlock_pkg['tested'].append(locked)
```
remove this line

vlours replied:
Line has been removed as suggested.

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,283 @@
+    state_pkg['present'] = []
+    state_pkg['missing'] = []
+    # For debug purpose you can include the listed packages in the debug
+    #state_pkg['listed'] = []
```
remove this line

vlours replied:
Line has been removed as suggested

gyptazy commented on library/yum_versionlock.py:
```diff
@@ -15,117 +15,283 @@
+            if pkg_version:
+                pkg_vers = pkg_name + '-' + pkg_version
+                # For debug purpose you can include the listed packages in the debug
+                #state_pkg['listed'].append(pkg_vers)
```
remove this line

vlours replied:
Lines have been removed as suggested

gyptazy reviewed (commented):
Hi,

thanks for the update. Looks fine, now. Maybe you should just remove the unused lines or add a dedicated debug mode to apply them. Removing them is also fine for me.

It's on you. Afterwards I merge it.

Thanks for your PR!

Regards,
Florian

vlours commented:
> Hi,
> 
> thanks for the update. Looks fine, now. Maybe you should just remove the unused lines or add a dedicated debug mode to apply them. Removing them is also fine for me.
> 
> It's on you. Afterwards I merge it.
> 
> Thanks for your PR!
> 
> Regards,
> Florian

Hi Florian,

Actually I've tried to get the debug messages when using the debug mode level '-vvvv'.
Unfortunately I failed and I was only able to get it at all verbose levels.

Do you know how to get the debug messages displayed only at the highest level of verbose ?

Cheers,
Vincent

Commit:
[versionlock] update of the versionlock python script
- Remove debug commented lines

vlours commented:
Hi Florian,

I've cleaned up the commented lines from my tests.

Cheers,
Vincent

florianpaulhoberg commented:
Hey @vlours,
thank you very much for your work and very fast changes on my notes!
This is merged, now.

Regards,
Florian
